File: internal/domain/artist/artist.go
package artist

type Artist struct {
	ID          int64
	Name        string
	Description string
	ImageURL    string
	Genres      []Genre
	Socials     []Social
}

func NewArtist(name, description, imageURL string, socials []Social, genres []Genre) (Artist, error) {
	return Artist{
		Name:        name,
		Description: description,
		ImageURL:    imageURL,
		Socials:     socials,
		Genres:      genres,
	}, nil
}

func (a *Artist) WithGenres(genres []Genre) {
	a.Genres = genres
}

func (a *Artist) WithSocials(socials []Social) {
	a.Socials = socials
}

File: frontend/src/lib/features/artist.ts
import { baseFields } from "$lib/model";
import { z } from "zod";

export const genreSchema = z.object({
	...baseFields.shape,
	name: z.string().nonempty()
});

export const socialSchema = z.object({
	...baseFields.shape,
	url: z.url()
});

export const artistSchema = z.object({
	...baseFields.shape,
	name: z.string().nonempty(),
	description: z.string().nonempty(),
	cover: z.url(),
	previewUrl: z.union([z.url(), z.literal("")]),
	genres: genreSchema.array().min(1),
	socials: socialSchema.array().optional()
});

export type Artist = z.infer<typeof artistSchema>;
export type Genre = z.infer<typeof genreSchema>;
export type Social = z.infer<typeof socialSchema>;
